index_impl.go

```diff
@@ -362,15 +366,76 @@ func (i *indexImpl) Search(req *SearchRequest) (sr *SearchResult, err error) {
 	return i.SearchInContext(context.Background(), req)
 }
 
+// memNeededForSearch is a helper function that returns an estimate of RAM
+// needed to execute a search request.
+func memNeededForSearch(req *SearchRequest,
+	searcher search.Searcher,
+	topnCollector *collector.TopNCollector) uint64 {
+
+	backingSize := req.Size + req.From + 1
+	if req.Size+req.From > collector.PreAllocSizeSkipCap {
+		backingSize = collector.PreAllocSizeSkipCap + 1
+	}
+	numDocMatches := backingSize + searcher.DocumentMatchPoolSize()
+
+	estimate := 0
+
+	// overhead, size in bytes from collector
+	estimate += topnCollector.Size()
+
+	var dm search.DocumentMatch
+	sizeOfDocumentMatch := dm.Size()
+
+	// pre-allocing DocumentMatchPool
+	var sc search.SearchContext
+	estimate += sc.Size() + numDocMatches*sizeOfDocumentMatch
+
+	// searcher overhead
+	estimate += searcher.Size()
+
+	// overhead from results, lowestMatchOutsideResults
+	estimate += (numDocMatches + 1) * sizeOfDocumentMatch
+
+	// additional overhead from SearchResult
+	var sr SearchResult
+	estimate += sr.Size()
+
+	// overhead from facet results
+	if req.Facets != nil {
+		var fr search.FacetResult
+		estimate += len(req.Facets) * fr.Size()
+	}
+
+	// highlighting, store
+	var d document.Document
+	if len(req.Fields) > 0 || req.Highlight != nil {
+		for i := 0; i < (req.Size + req.From); i++ { // size + from => number of hits
+			estimate += (req.Size + req.From) * d.Size()
+		}
+	}
+
+	return uint64(estimate)
+}
+
 // SearchInContext executes a search request operation within the provided
-// Context.  Returns a SearchResult object or an error.
+// Context. Returns a SearchResult object or an error.
 func (i *indexImpl) SearchInContext(ctx context.Context, req *SearchRequest) (sr *SearchResult, err error) {
 	i.mutex.RLock()
 	defer i.mutex.RUnlock()
 
+	memCheckCallback := func(size uint64) error {
+		if memCb := ctx.Value(SearchMemCheckCallbackKey); memCb != nil {
+			if memCbFn, ok := memCb.(SearchMemCheckCallbackFn); ok {
+				return memCbFn(size)
+			}
+		}
+		return nil
+	}
+
 	searchStart := time.Now()
 
 	if !i.open {
+		_ = memCheckCallback(0)
```

Comment:
what is the purpose of these memCheckCallback(0) calls? there are several of them

Reply:
The client (that sets up the callback function) will be responsible for aggregating the estimates from all the local pindexes before responding with a decision. This is just to make sure that every pindex invokes the callback that one time regardless of an error situation.

Reply:
OK, again I don't think that is quite the right design.  I don't think the estimate of the coordinating query should include the local queries.  Executing each individual query has some requirements, and merging them has requirements, we can (and for simplicity) treat these as independent.  The only relation they have is that estimating the size of the merge requirements is also based on the SearchRequest.  But it is a different estimate, and does not include the child queries memory requirements.

In order to get this merged, let's just simplify further.  SearcInContext invokes the callback it it's present in the context.  And we don't call it otherwise.  Callers have no guarantee that the callback will ever be invoked.

Reply:
Ah understood, thanks for that. I'll have patch up in a sec